app/server/lib/ActiveDoc.ts

```diff
@@ -2337,6 +2359,9 @@ export class ActiveDoc extends EventEmitter {
     try {
 
       this.setMuted();
+      // No timeout on this callback: if it hangs, it will make the document unusable.
+      await options.beforeShutdown?.();
```

Comment:
Mmm if there's no reason against it (any maybe there is), I'd put this after disabling the inactivity timer. Shouldn't matter, just nice to have that timer off synchronously so as not to have to think about it.

Reply:
Done in https://github.com/gristlabs/grist-core/pull/1624/commits/cc81a1d5612756bb4f303bb9e775d56a03116cd9